### PR TITLE
[HOTPLUG] Add device properties dialog support

### PR DESCRIPTION
**Pull request by TAN-Gaming**
## Proposed changes
- Clicking at "Properties" button or right-click menu now opens properties dialog of the selected device

VirtualBox_ReactOS_22_10_2023_02_35_34

### CODE REVIEW

File: dll/cpl/hotplug/hotplug.c [hunk 1]
```diff
@@ -344,6 +344,78 @@ ShowContextMenu(
...
+    FreeLibrary(hDll);
+
+    return ret;
+}
+
+static
+VOID
+ShowDeviceProperties(
+    _In_ HWND hwndParent,
+    _In_ DWORD DevInst)
```
Comment (HBelusca) [inline dll/cpl/hotplug/hotplug.c:398]: This should be a `DEVINST` parameter. (If using that type makes compilaton fail, then you may need to include `cfgmgr32.h` .)

EDIT: I see that the existing code (above in that file) uses DWORD, but let's use the actual correct types when possible.

File: dll/cpl/hotplug/hotplug.c [hunk 2]
```diff
@@ -344,6 +344,78 @@ ShowContextMenu(
...
+{
+    ULONG ulSize;
+    CONFIGRET cr;
+    LPWSTR pszDevId;
+
+    cr = CM_Get_Device_ID_Size(&ulSize, DevInst, 0);
+    if (cr != CR_SUCCESS || ulSize == 0)
+        return;
+
+    pszDevId = HeapAlloc(GetProcessHeap(), 0, (ulSize + 1) * sizeof(WCHAR));
```
Comment (HBelusca) [inline dll/cpl/hotplug/hotplug.c:408]:
```suggestion
    /* Take the terminating NULL into account */
    ulSize++;
    pszDevId = HeapAlloc(GetProcessHeap(), 0, ulSize * sizeof(WCHAR));
```

File: dll/cpl/hotplug/hotplug.c [hunk 3]
```diff
@@ -344,6 +344,78 @@ ShowContextMenu(
...
+    if (pszDevId == NULL)
+        return;
+
+    cr = CM_Get_Device_IDW(DevInst, pszDevId, ulSize + 1, 0);
```
Comment (HBelusca) [inline dll/cpl/hotplug/hotplug.c:412]:
```suggestion
    cr = CM_Get_Device_IDW(DevInst, pszDevId, ulSize, 0);
```

File: dll/cpl/hotplug/hotplug.c [hunk 4]
```diff
@@ -344,6 +344,78 @@ ShowContextMenu(
...
+
+    return item.lParam;
+}
+
+static
+int
+_DevicePropertiesW(
+    _In_ HWND hWndParent,
+    _In_ LPCWSTR lpMachineName,
+    _In_ LPCWSTR lpDeviceID,
```
Comment (HBelusca) [inline dll/cpl/hotplug/hotplug.c:373]:
```suggestion
    _In_opt_ HWND hWndParent,
    _In_opt_ LPCWSTR lpMachineName,
    _In_opt_ LPCWSTR lpDeviceID,
```
In principle all these parameters are optional (by checking the implementation of DeviceProperties(Ex)W)

Comment (HBelusca) [reply dll/cpl/hotplug/hotplug.c:373]: In addition, I don't know whether you plan to use this DevicePropertiesW with other sorts of parameters later on, but if not, you may instead consider simplifying its implementation, much like what is done here for example:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/shellext/deskadp/deskadp.c;hb=855008d97b2c7e1a1b72ad268b68d83c739ebc6d#l267

File: dll/cpl/hotplug/hotplug.c (hunk 1 above)
Comment (TAN-Gaming) [reply dll/cpl/hotplug/hotplug.c]: Done.

File: dll/cpl/hotplug/hotplug.c (hunk 4 above)
Comment (TAN-Gaming) [reply dll/cpl/hotplug/hotplug.c]: Function removed.

File: dll/cpl/hotplug/hotplug.c (hunk 2 above)
Comment (TAN-Gaming) [reply dll/cpl/hotplug/hotplug.c]: Done.

File: dll/cpl/hotplug/hotplug.c (hunk 3 above)
Comment (TAN-Gaming) [reply dll/cpl/hotplug/hotplug.c]: Done.

File: dll/cpl/hotplug/hotplug.c [hunk 5]
```diff
@@ -344,6 +345,68 @@ ShowContextMenu(
...
+        return;
+
+    cr = CM_Get_Device_IDW(DevInst, pszDevId, ulSize, 0);
+    if (cr == CR_SUCCESS)
+    {
+        HMODULE hDevMgrDll;
+        PFDEVICEPROPERTIESW pDevicePropertiesW;
+
+        hDevMgrDll = LoadLibraryW(L"devmgr.dll");
+        if (hDevMgrDll != NULL)
```
Comment (HBelusca) [inline dll/cpl/hotplug/hotplug.c:397]: Optional changes:
```suggestion
        if (hDevMgrDll)
```

File: dll/cpl/hotplug/hotplug.c [hunk 6]
```diff
@@ -344,6 +345,68 @@ ShowContextMenu(
...
+        {
+            pDevicePropertiesW = (PFDEVICEPROPERTIESW)GetProcAddress(hDevMgrDll, "DevicePropertiesW");
+            if (pDevicePropertiesW != NULL)
```
Comment (HBelusca) [inline dll/cpl/hotplug/hotplug.c:400]:
```suggestion
            if (pDevicePropertiesW)
```

File: dll/cpl/hotplug/hotplug.c (hunk 5 above)
Comment (TAN-Gaming) [reply dll/cpl/hotplug/hotplug.c:397]: I follow the original style.

File: dll/cpl/hotplug/hotplug.c [hunk 7]
```diff
@@ -24,6 +24,7 @@ typedef struct _HOTPLUG_DATA
     DWORD dwFlags;
 } HOTPLUG_DATA, *PHOTPLUG_DATA;
 
+typedef int (WINAPI *PFDEVICEPROPERTIESW)(HWND, LPCWSTR, LPCWSTR, BOOL);
```
[the remark this answers is not in the document]
Comment (TAN-Gaming) [reply dll/cpl/hotplug/hotplug.c:27]: OK.

**Review by HBelusca — APPROVED** [review]

**Review by binarymaster — APPROVED** [review]
LGTM. 👍 You may want to squash commits into single one.

**Pull request merged**